= bamstats

https://travis-ci.org/guigolab/bamstats[image:https://travis-ci.org/guigolab/bamstats.svg?branch=develop["Build Status"]^]
https://coveralls.io/github/guigolab/bamstats[image:https://coveralls.io/repos/github/guigolab/bamstats/badge.svg?branch=develop[Coverage Status]]

`Bamstats` is a command line tool written in `Go` for computing mapping statistics from a `BAM` file.

It currently supports the following statistics:

- general
- genome coverage

== General statistics

The general mapping statistics include:

- Total number of reads
- Number of unmapped reads
- Number of mapped reads grouped by number of multimaps (`NH` tag in `BAM` file)
- Number of mappings
- Ratio of mappings vs mapped reads

If the data is paired-end, a section for read-pairs is also reported. In addition to the above metrics, the section contains a map of the insert size length and the corresponding support as number of reads.

== Genome coverage statistics

The genome coverage ststistics are computed for RNA-seq data and include counts for the following genomic regions:

- exon
- intron
- exonic_intronic
- intergenic
- others

The above metrics are computed for continuous and split mapped reads. An aggregated total is computed across elements and read types too.

The `--uniq` (or `-u`) command line flag allows reporting of genome coverage statistics for uniquely mapped reads too.

== Output file example:

An example of the output `JSON` file with all the statistics is shown below. The `insert_sizes` map has been truncated for readability.

[source, json]
----
{
	"coverage": {
		"total": {
			"exonic_intronic": 4101,
			"intron": 24977,
			"exon": 13058,
			"intergenic": 4,
			"others": 0,
			"total": 42140
		},
		"continuous": {
			"exonic_intronic": 3700,
			"intron": 24905,
			"exon": 6882,
			"intergenic": 4,
			"others": 0,
			"total": 35491
		},
		"split": {
			"exonic_intronic": 401,
			"intron": 72,
			"exon": 6176,
			"intergenic": 0,
			"others": 0,
			"total": 6649
		}
	},
	"general": {
		"reads": {
			"total": 42140,
			"mapped": {
				"1": 13639,
				"2": 7867,
				"3": 7769,
				"4": 5934,
				"5": 3876,
				"6": 1838,
				"7": 1016,
				"8": 171,
				"9": 28,
				"10": 2
			},
			"mappings": {
				"ratio": 2.3730422401518747,
				"count": 100000
			}
		},
		"pairs": {
			"total": 21649,
			"mapped": {
				"1": 6946,
				"2": 4023,
				"3": 3924,
				"4": 3136,
				"5": 2060,
				"6": 940,
				"7": 516,
				"8": 89,
				"9": 14,
				"10": 1
			},
			"insert_sizes": {
				...
				"100": 56,
				"101": 59,
				"102": 71,
				"103": 52,
				"104": 52,
				"105": 54,
				"106": 75,
				"107": 58,
				"108": 75,
				"109": 64,
				"110": 70,
				"111": 102,
				"112": 86,
				"113": 85,
				"114": 89,
				"115": 84,
				"116": 138,
				"117": 102,
				"118": 137,
				"119": 94,
				"120": 142,
				"121": 127,
				"122": 127,
				"123": 120,
				"124": 153,
				"125": 134,
				"126": 167,
				"127": 148,
				"128": 135,
				"129": 159,
				"130": 207,
				"131": 133,
				"132": 152,
				"133": 158,
				"134": 159,
				"135": 202,
				"136": 191,
				"137": 171,
				"138": 183,
				"139": 184,
				"140": 204,
				"141": 160,
				"142": 161,
				"143": 159,
				"144": 174,
				"145": 161,
				"146": 130,
				"147": 115,
				"148": 192,
				"149": 209,
				"150": 186,
				"151": 141,
				"152": 137,
				"153": 144,
				"154": 129,
				"155": 142,
				"156": 116,
				"157": 131,
				"158": 148,
				"159": 167,
				"160": 134,
				"161": 116,
				"162": 121,
				"163": 131,
				"164": 138,
				"165": 135,
				"166": 106,
				"167": 101,
				"168": 122,
				"169": 120,
				"170": 124,
				"171": 113,
				"172": 124,
				"173": 114,
				"174": 139,
				"175": 110,
				"176": 135,
				"177": 112,
				"178": 86,
				"179": 90,
				"180": 97,
				"181": 97,
				"182": 105,
				"183": 73,
				"184": 130,
				"185": 106,
				"186": 91,
				"187": 90,
				"188": 98,
				"189": 83,
				"190": 83,
				"191": 74,
				"192": 81,
				"193": 75,
				"194": 77,
				"195": 86,
				"196": 96,
				"197": 96,
				"198": 75,
				"199": 79,
				"200": 84,
				"201": 82,
				"202": 59,
				"203": 53,
				"204": 61,
				"205": 62,
				"206": 61,
				"207": 80,
				"208": 63,
				"209": 65,
				"210": 52,
				"211": 60,
				"212": 68,
				"213": 66,
				"214": 54,
				"215": 62,
				"216": 53,
				"217": 51,
				"218": 58,
				"219": 55,
				"220": 50,
				"221": 51,
				...
			}
		}
	}
}
----

The genomic coverage statistics for uniquely mapped reads will be reported as an additional JSON object:

[source, json]
----
{
	"coverage": {
		"total": {
			"exonic_intronic": 4101,
			"intron": 24977,
			"exon": 13058,
			"intergenic": 4,
			"others": 0,
			"total": 42140
		},
		"continuous": {
			"exonic_intronic": 3700,
			"intron": 24905,
			"exon": 6882,
			"intergenic": 4,
			"others": 0,
			"total": 35491
		},
		"split": {
			"exonic_intronic": 401,
			"intron": 72,
			"exon": 6176,
			"intergenic": 0,
			"others": 0,
			"total": 6649
		}
	},
	"coverageUniq": {
		"total": {
			"exonic_intronic": 1034,
			"intron": 10751,
			"exon": 1850,
			"intergenic": 4,
			"others": 0,
			"total": 13639
		},
		"continuous": {
			"exonic_intronic": 933,
			"intron": 10682,
			"exon": 1083,
			"intergenic": 4,
			"others": 0,
			"total": 12702
		},
		"split": {
			"exonic_intronic": 101,
			"intron": 69,
			"exon": 767,
			"intergenic": 0,
			"others": 0,
			"total": 937
		}
	},
	"general": {
	...
----

== License
This software is release under a BSD-style license that can be found in the `LICENSE` file.
